azext_iot/models/parsers.py

```diff
@@ -0,0 +1,248 @@
+        if simulate_errors and i == 1:
+            create_encoding_error = True
+        if simulate_errors and i == 2:
+            create_custom_header_warning = True
+        if simulate_errors and i == 3:
+            create_payload_error = True
+
+        system_properties = self._parse_system_properties(message)
+
+        self._parse_content_encoding(message, system_properties, create_encoding_error)
+
+        event = {}
+
+        origin_device_id = self.parse_device_id(message)
+        event["origin"] = origin_device_id
+
+        self._parse_content_type(
+            content_type_hint,
+            system_properties,
+            origin_device_id,
+            create_custom_header_warning,
+        )
+
+        if pnp_context:
+            message_interface_name = self._parse_interface_name(
+                message, pnp_context, interface_name, origin_device_id
+            )
+
+            event["interface"] = message_interface_name
+
+        if properties:
+            event["properties"] = {}
+
+        if "anno" in properties or "all" in properties:
+            annotations = self._parse_annotations(message)
+            event["properties"]["annotations"] = annotations
+
+        if system_properties and ("sys" in properties or "all" in properties):
+            event["properties"]["system"] = system_properties
+
+        if "app" in properties or "all" in properties:
+            application_properties = self._parse_application_properties(message)
+            event["properties"]["application"] = application_properties
+
+        payload = self._parse_payload(message, origin_device_id, create_payload_error)
+        if not payload:
+            return {}
+
+        event["payload"] = payload
+
+        event_source = {"event": event}
+
+        return event_source
+
+    def parse_device_id(self, message: Message) -> str:
+        try:
+            return str(message.annotations.get(DEVICE_ID_IDENTIFIER), "utf8")
+        except Exception:
+            self._errors.append(f"Device id not found in message: {message}")
+
+    def write_logs(self) -> None:
+        for error in self._errors:
+            self._logger.error("[Error] " + error)
+
+        for warning in self._warnings:
+            self._logger.warn("[Warning] " + warning)
+
+        for info in self._info:
+            self._logger.info("[Info] " + info)
+
+    def _reset_issues(self) -> None:
+        self._info = []
+        self._warnings = []
+        self._errors = []
+
+    def _parse_interface_name(
+        self, message: Message, pnp_context, interface_name, origin_device_id
+    ) -> str:
+        message_interface_name = ""
+
+        try:
+            message_interface_name = str(
+                message.annotations.get(INTERFACE_NAME_IDENTIFIER), "utf8"
+            )
+        except Exception:
+            self._errors.append(
+                f"Unable to parse interface_name given a pnp_device. {origin_device_id}. "
```

Comment:
Can you please verify if f strings work in python 3.5? Last I recall it was min python 3.6. We have to support at least python 3.5.

Reply:
Ah good catch, they are only available in 3.6 and up. I might update the contribution docs to suggest spinning up the env using python 3.5 or below for backwards compat